### ISSUE TITLE
"Instruction does not dominate all uses" error with shufflevector and icmp

### ISSUE THREAD

**Issue by vient** [llvm/llvm-project#64929]
Compiling LLVM 17.0.0-rc3 with full LTO and PGO, got the following error
```console
Instruction does not dominate all uses!
  %157 = icmp eq <2 x ptr> %156, %138
  %155 = shufflevector <2 x i1> %157, <2 x i1> %154, <4 x i32> <i32 1, i32 0, i32 3, i32 2>
LLVM ERROR: Broken module found, compilation aborted!
```

Link commands look like this
```shell
/work/stage1/install/bin/clang++ -march=x86-64-v3 -fno-pie -gdwarf-4 -g1 -Wno-backend-plugin -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fprofile-instr-use="/work/stage2-prof-gen/profiles/clang.profdata" -flto=full -O3 -DNDEBUG -Wl,--emit-relocs -no-pie -fuse-ld=lld -fuse-ld=lld -Wl,--color-diagnostics -fprofile-instr-use="/work/stage2-prof-gen/profiles/clang.profdata" -flto=full    -Wl,--gc-sections tools/dsymutil/CMakeFiles/dsymutil.dir/dsymutil.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/BinaryHolder.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/DebugMap.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/DwarfLinkerForBinary.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/MachODebugMapParser.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/MachOUtils.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/Reproducer.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/SymbolMap.cpp.o -o bin/dsymutil  -Wl,-rpath,"\$ORIGIN/../lib:"  lib/libLLVMX86CodeGen.a  lib/libLLVMX86Desc.a  lib/libLLVMX86Info.a  lib/libLLVMAsmPrinter.a  lib/libLLVMCodeGen.a  lib/libLLVMCodeGenTypes.a  lib/libLLVMDWARFLinker.a  lib/libLLVMDWARFLinkerParallel.a  lib/libLLVMDebugInfoDWARF.a  lib/libLLVMMC.a  lib/libLLVMObject.a  lib/libLLVMOption.a  lib/libLLVMRemarks.a  lib/libLLVMSupport.a  lib/libLLVMTarget.a  lib/libLLVMTargetParser.a  -lpthread  lib/libLLVMMCDisassembler.a  lib/libLLVMCFGuard.a  lib/libLLVMGlobalISel.a  lib/libLLVMInstrumentation.a  lib/libLLVMSelectionDAG.a  lib/libLLVMAsmPrinter.a  lib/libLLVMCodeGen.a  lib/libLLVMCodeGenTypes.a  lib/libLLVMTarget.a  lib/libLLVMBitWriter.a  lib/libLLVMObjCARCOpts.a  lib/libLLVMScalarOpts.a  lib/libLLVMAggressiveInstCombine.a  lib/libLLVMInstCombine.a  lib/libLLVMTransformUtils.a  lib/libLLVMAnalysis.a  lib/libLLVMProfileData.a  lib/libLLVMSymbolize.a  lib/libLLVMDebugInfoPDB.a  lib/libLLVMDebugInfoMSF.a  lib/libLLVMDebugInfoBTF.a  lib/libLLVMDebugInfoDWARF.a  lib/libLLVMObject.a  lib/libLLVMMCParser.a  lib/libLLVMMC.a  lib/libLLVMDebugInfoCodeView.a  lib/libLLVMIRReader.a  lib/libLLVMBitReader.a  lib/libLLVMAsmParser.a  lib/libLLVMCore.a  lib/libLLVMRemarks.a  lib/libLLVMBitstreamReader.a  lib/libLLVMTextAPI.a  lib/libLLVMBinaryFormat.a  lib/libLLVMTargetParser.a  lib/libLLVMSupport.a  -lrt  -ldl  -lpthread  -lm  /usr/lib/x86_64-linux-gnu/libz.so  lib/libLLVMDemangle.a
```

I assume it may be better to provide you with a reproducer? I don't have it at the moment since it occurred in CI.

**Comment by fhahn** [comment]
> I assume it may be better to provide you with a reproducer? 

Yes, unfortunately without a reproducer there's not much we can do to fix that, especially the build above uses PGO.

**Comment by vient** [comment]
Here is the reproducer that worked on my system. I compile clang 17.0.0-rc3, compile profiled version, gather profiles on clang itself and then try to use that profile.

[clang.profdata used in the last PGO build](https://github.com/llvm/llvm-project/files/12508377/clang.profdata.zip)

Script itself
```bash
#!/bin/bash -eux

export LLD_IN_TEST=1

function assert_not_empty {
    [ -n "$1" ]
}

LLVM_GIT_BRANCH="llvmorg-17.0.0-rc3"

DEBUG_FLAGS="-gdwarf-4 -g1"
LLVM_C_FLAGS="-march=x86-64-v3 -fno-pie ${DEBUG_FLAGS}"
LLVM_CXX_FLAGS="${LLVM_C_FLAGS}"
LLVM_SHARED_LINKER_FLAGS="-Wl,--emit-relocs"
LLVM_EXE_LINKER_FLAGS="${LLVM_SHARED_LINKER_FLAGS} -no-pie"

TOPLEV="$(realpath work)"
LLVM_PROJECT_DIR="${TOPLEV}/llvm-project"
BASE_COMPILER_DIR="${TOPLEV}/stage1"
BINUTILS_INCLUDE_DIR="${TOPLEV}/binutils/include"

mkdir -p "${TOPLEV}"
cd "$_"

git clone --depth 1 git://sourceware.org/git/binutils-gdb.git binutils &
git clone -b "${LLVM_GIT_BRANCH}" --depth=1 https://github.com/llvm/llvm-project.git &

sudo apt-get upgrade --yes --with-new-pkgs
sudo apt-get install --yes \
    apt-transport-https \
    build-essential \
    cmake \
    coreutils \
    curl \
    git \
    libc6-dev-i386 \
    linux-tools-generic \
    psmisc \
    python3 \
    python3-pip \
    sed \
    software-properties-common \
    vim \
    wget

python3 -m pip install ninja

wait

mkdir -p "${BASE_COMPILER_DIR}"
cd "$_"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_FLAGS="${LLVM_C_FLAGS}" \
    -DCMAKE_CXX_FLAGS="${LLVM_CXX_FLAGS}" \
    -DCMAKE_EXE_LINKER_FLAGS="-no-pie" \
    -DCMAKE_INSTALL_PREFIX="${BASE_COMPILER_DIR}/install" \
    -DCOMPILER_RT_BUILD_LIBFUZZER=OFF \
    -DCOMPILER_RT_BUILD_SANITIZERS=OFF \
    -DCOMPILER_RT_BUILD_XRAY=OFF \
    -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld;bolt" \
    -DLLVM_ENABLE_RUNTIMES="compiler-rt" \
    -DLLVM_TARGETS_TO_BUILD=X86 \
    -DLLVM_USE_LINKER=lld
ninja install
ninja -t clean

# build self-profiling compiler
mkdir -p "${TOPLEV}/stage2-prof-gen"
cd "$_"
CPATH="${BASE_COMPILER_DIR}/install/bin"
LLVM_EXE_LINKER_FLAGS="${LLVM_EXE_LINKER_FLAGS} -fuse-ld=lld"
LLVM_SHARED_LINKER_FLAGS="${LLVM_SHARED_LINKER_FLAGS} -fuse-ld=lld"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_COMPILER="${CPATH}/clang" \
    -DCMAKE_C_FLAGS="${LLVM_C_FLAGS} -Xclang -mllvm -Xclang -vp-counters-per-site=256" \
    -DCMAKE_CXX_COMPILER="${CPATH}/clang++" \
    -DCMAKE_CXX_FLAGS="${LLVM_CXX_FLAGS} -Xclang -mllvm -Xclang -vp-counters-per-site=256" \
    -DCMAKE_EXE_LINKER_FLAGS="${LLVM_EXE_LINKER_FLAGS}" \
    -DCMAKE_SHARED_LINKER_FLAGS="${LLVM_SHARED_LINKER_FLAGS}" \
    -DCMAKE_INSTALL_PREFIX="${TOPLEV}/stage2-prof-gen/install" \
    -DLLVM_BINUTILS_INCDIR="${BINUTILS_INCLUDE_DIR}" \
    -DLLVM_BUILD_INSTRUMENTED=IR \
    -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld" \
    -DLLVM_ENABLE_RUNTIMES="compiler-rt" \
    -DLLVM_ENABLE_TERMINFO=OFF \
    -DLLVM_PROFILE_DATA_DIR="${TOPLEV}/stage2-prof-gen/profiles/%p" \
    -DLLVM_TARGETS_TO_BUILD=X86 \
    -DLLVM_USE_LINKER=lld
ninja install
ninja -t clean

# gather compiler profiles
CPATH="${TOPLEV}/stage2-prof-gen/install/bin"

mkdir -p "${TOPLEV}/stage3-train-llvm-release"
cd "$_"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_COMPILER="${CPATH}/clang" \
    -DCMAKE_C_FLAGS="${LLVM_C_FLAGS}" \
    -DCMAKE_CXX_COMPILER="${CPATH}/clang++" \
    -DCMAKE_CXX_FLAGS="${LLVM_CXX_FLAGS}" \
    -DCMAKE_EXE_LINKER_FLAGS="-no-pie" \
    -DCMAKE_INSTALL_PREFIX="${PWD}/install" \
    -DCOMPILER_RT_BUILD_LIBFUZZER=OFF \
    -DCOMPILER_RT_BUILD_SANITIZERS=OFF \
    -DCOMPILER_RT_BUILD_XRAY=OFF \
    -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld;bolt" \
    -DLLVM_ENABLE_RUNTIMES="compiler-rt" \
    -DLLVM_TARGETS_TO_BUILD=X86 \
    -DLLVM_USE_LINKER=lld
ninja

rm -rf "${TOPLEV}/stage3-train"*

# merge profiles
cd "${TOPLEV}/stage2-prof-gen/profiles"
"${BASE_COMPILER_DIR}/install/bin/llvm-profdata" merge -output=clang.profdata ./*

# build PGO compiler
mkdir -p "${TOPLEV}/stage2-prof-use-lto"
cd "$_"
CPATH="${BASE_COMPILER_DIR}/install/bin"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_COMPILER="${CPATH}/clang" \
    -DCMAKE_C_FLAGS="${LLVM_C_FLAGS} -Wno-backend-plugin" \
    -DCMAKE_CXX_COMPILER="${CPATH}/clang++" \
    -DCMAKE_CXX_FLAGS="${LLVM_CXX_FLAGS} -Wno-backend-plugin" \
    -DCMAKE_EXE_LINKER_FLAGS="${LLVM_EXE_LINKER_FLAGS}" \
    -DCMAKE_INSTALL_PREFIX="${TOPLEV}/stage2-prof-use-lto/install" \
    -DCMAKE_SHARED_LINKER_FLAGS="${LLVM_SHARED_LINKER_FLAGS}" \
    -DLLVM_BINUTILS_INCDIR="${BINUTILS_INCLUDE_DIR}" \
    -DLLVM_ENABLE_LTO=Full \
    -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld;bolt" \
    -DLLVM_ENABLE_RUNTIMES="compiler-rt" \
    -DLLVM_PARALLEL_LINK_JOBS=30 \
    -DLLVM_PROFDATA_FILE="${TOPLEV}/stage2-prof-gen/profiles/clang.profdata" \
    -DLLVM_TARGETS_TO_BUILD=X86 \
    -DLLVM_USE_LINKER=lld
ninja bin/dsymutil  # error
```

Error
```
/twix/packages/twix-clang/work/stage1/install/bin/clang++ -Wno-backend-plugin -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fprofile-instr-use="/twix/packages/twix-clang/work/stage2-prof-gen/profiles/clang.profdata" -flto=full -O3 -DNDEBUG -fuse-ld=lld -Wl,--color-diagnostics -fprofile-instr-use="/twix/packages/twix-clang/work/stage2-prof-gen/profiles/clang.profdata" -flto=full    -Wl,--gc-sections tools/dsymutil/CMakeFiles/dsymutil.dir/dsymutil.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/BinaryHolder.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/DebugMap.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/DwarfLinkerForBinary.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/MachODebugMapParser.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/MachOUtils.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/Reproducer.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/SymbolMap.cpp.o -o bin/dsymutil  -Wl,-rpath,"\$ORIGIN/../lib:"  lib/libLLVMX86CodeGen.a  lib/libLLVMX86Desc.a  lib/libLLVMX86Info.a  lib/libLLVMAsmPrinter.a  lib/libLLVMCodeGen.a  lib/libLLVMCodeGenTypes.a  lib/libLLVMDWARFLinker.a  lib/libLLVMDWARFLinkerParallel.a  lib/libLLVMDebugInfoDWARF.a  lib/libLLVMMC.a  lib/libLLVMObject.a  lib/libLLVMOption.a  lib/libLLVMRemarks.a  lib/libLLVMSupport.a  lib/libLLVMTarget.a  lib/libLLVMTargetParser.a  -lpthread  lib/libLLVMMCDisassembler.a  lib/libLLVMCFGuard.a  lib/libLLVMGlobalISel.a  lib/libLLVMInstrumentation.a  lib/libLLVMSelectionDAG.a  lib/libLLVMAsmPrinter.a  lib/libLLVMCodeGen.a  lib/libLLVMCodeGenTypes.a  lib/libLLVMTarget.a  lib/libLLVMBitWriter.a  lib/libLLVMObjCARCOpts.a  lib/libLLVMScalarOpts.a  lib/libLLVMAggressiveInstCombine.a  lib/libLLVMInstCombine.a  lib/libLLVMTransformUtils.a  lib/libLLVMAnalysis.a  lib/libLLVMProfileData.a  lib/libLLVMSymbolize.a  lib/libLLVMDebugInfoPDB.a  lib/libLLVMDebugInfoMSF.a  lib/libLLVMDebugInfoBTF.a  lib/libLLVMDebugInfoDWARF.a  lib/libLLVMObject.a  lib/libLLVMMCParser.a  lib/libLLVMMC.a  lib/libLLVMDebugInfoCodeView.a  lib/libLLVMIRReader.a  lib/libLLVMBitReader.a  lib/libLLVMAsmParser.a  lib/libLLVMCore.a  lib/libLLVMRemarks.a  lib/libLLVMBitstreamReader.a  lib/libLLVMTextAPI.a  lib/libLLVMBinaryFormat.a  lib/libLLVMTargetParser.a  lib/libLLVMSupport.a  -lrt  -ldl  -lpthread  -lm  /usr/lib/x86_64-linux-gnu/libz.so  lib/libLLVMDemangle.a && :
Instruction does not dominate all uses!
  %157 = icmp eq <2 x ptr> %156, %138
  %155 = shufflevector <2 x i1> %157, <2 x i1> %154, <4 x i32> <i32 1, i32 0, i32 3, i32 2>
LLVM ERROR: Broken module found, compilation aborted!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /twix/packages/twix-clang/work/stage1/install/bin/ld.lld -pie -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o bin/dsymutil /lib/x86_64-linux-gnu/Scrt1.o /lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/13.2.0/crtbeginS.o -L/twix/packages/twix-clang/work/stage1/install/lib/clang/17/lib/x86_64-unknown-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/13.2.0 -L/usr/lib/gcc/x86_64-linux-gnu/13.2.0/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/lib -L/usr/lib -plugin-opt=mcpu=x86-64 -plugin-opt=O3 -plugin-opt=-function-sections=1 -plugin-opt=-data-sections=1 -plugin-opt=cs-profile-path=/twix/packages/twix-clang/work/stage2-prof-gen/profiles/clang.profdata --color-diagnostics --gc-sections tools/dsymutil/CMakeFiles/dsymutil.dir/dsymutil.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/BinaryHolder.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/CFBundle.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/DebugMap.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/DwarfLinkerForBinary.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/MachODebugMapParser.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/MachOUtils.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/Reproducer.cpp.o tools/dsymutil/CMakeFiles/dsymutil.dir/SymbolMap.cpp.o -rpath $ORIGIN/../lib: lib/libLLVMX86CodeGen.a lib/libLLVMX86Desc.a lib/libLLVMX86Info.a lib/libLLVMAsmPrinter.a lib/libLLVMCodeGen.a lib/libLLVMCodeGenTypes.a lib/libLLVMDWARFLinker.a lib/libLLVMDWARFLinkerParallel.a lib/libLLVMDebugInfoDWARF.a lib/libLLVMMC.a lib/libLLVMObject.a lib/libLLVMOption.a lib/libLLVMRemarks.a lib/libLLVMSupport.a lib/libLLVMTarget.a lib/libLLVMTargetParser.a -lpthread lib/libLLVMMCDisassembler.a lib/libLLVMCFGuard.a lib/libLLVMGlobalISel.a lib/libLLVMInstrumentation.a lib/libLLVMSelectionDAG.a lib/libLLVMAsmPrinter.a lib/libLLVMCodeGen.a lib/libLLVMCodeGenTypes.a lib/libLLVMTarget.a lib/libLLVMBitWriter.a lib/libLLVMObjCARCOpts.a lib/libLLVMScalarOpts.a lib/libLLVMAggressiveInstCombine.a lib/libLLVMInstCombine.a lib/libLLVMTransformUtils.a lib/libLLVMAnalysis.a lib/libLLVMProfileData.a lib/libLLVMSymbolize.a lib/libLLVMDebugInfoPDB.a lib/libLLVMDebugInfoMSF.a lib/libLLVMDebugInfoBTF.a lib/libLLVMDebugInfoDWARF.a lib/libLLVMObject.a lib/libLLVMMCParser.a lib/libLLVMMC.a lib/libLLVMDebugInfoCodeView.a lib/libLLVMIRReader.a lib/libLLVMBitReader.a lib/libLLVMAsmParser.a lib/libLLVMCore.a lib/libLLVMRemarks.a lib/libLLVMBitstreamReader.a lib/libLLVMTextAPI.a lib/libLLVMBinaryFormat.a lib/libLLVMTargetParser.a lib/libLLVMSupport.a -lrt -ldl -lpthread -lm /usr/lib/x86_64-linux-gnu/libz.so lib/libLLVMDemangle.a -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/13.2.0/crtendS.o /lib/x86_64-linux-gnu/crtn.o
 #0 0x0000000000fcb12b llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /twix/packages/twix-clang/work/llvm-project/llvm/lib/Support/Unix/Signals.inc:602:22
 #1 0x0000000000fc892b llvm::sys::RunSignalHandlers() /twix/packages/twix-clang/work/llvm-project/llvm/lib/Support/Signals.cpp:104:20
 #2 0x0000000000fc892b SignalHandler(int) /twix/packages/twix-clang/work/llvm-project/llvm/lib/Support/Unix/Signals.inc:403:31
 #3 0x00007fc94cfdc420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #4 0x00007fc94ca1300b raise /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007fc94c9f2859 abort /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:81:7
 #6 0x0000000000f322d6 (/twix/packages/twix-clang/work/stage1/install/bin/ld.lld+0xf322d6)
 #7 0x0000000000f324b8 (/twix/packages/twix-clang/work/stage1/install/bin/ld.lld+0xf324b8)
 #8 0x0000000003a515db (/twix/packages/twix-clang/work/stage1/install/bin/ld.lld+0x3a515db)
 #9 0x0000000001d46348 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /twix/packages/twix-clang/work/llvm-project/llvm/include/llvm/IR/PassManager.h:521:20
#10 0x0000000001d46348 runNewPMPasses(llvm::lto::Config const&, llvm::Module&, llvm::TargetMachine*, unsigned int, bool, llvm::ModuleSummaryIndex*, llvm::ModuleSummaryIndex const*) /twix/packages/twix-clang/work/llvm-project/llvm/lib/LTO/LTOBackend.cpp:335:10
#11 0x0000000001d49df2 llvm::lto::opt(llvm::lto::Config const&, llvm::TargetMachine*, unsigned int, llvm::Module&, bool, llvm::ModuleSummaryIndex*, llvm::ModuleSummaryIndex const*, std::vector<unsigned char, std::allocator<unsigned char>> const&) /twix/packages/twix-clang/work/llvm-project/llvm/lib/LTO/LTOBackend.cpp:362:34
#12 0x0000000001d49df2 llvm::lto::backend(llvm::lto::Config const&, std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, unsigned int, llvm::Module&, llvm::ModuleSummaryIndex&) /twix/packages/twix-clang/work/llvm-project/llvm/lib/LTO/LTOBackend.cpp:510:13
#13 0x0000000001d3ba21 std::_Function_base::~_Function_base() /usr/local/twix-gcc-13/include/c++/13.2.0/bits/std_function.h:243:11
#14 0x0000000001d3ba21 std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>::~function() /usr/local/twix-gcc-13/include/c++/13.2.0/bits/std_function.h:334:11
#15 0x0000000001d3ba21 llvm::lto::LTO::runRegularLTO(std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>) /twix/packages/twix-clang/work/llvm-project/llvm/lib/LTO/LTO.cpp:1343:70
#16 0x0000000001d3bf93 std::_Function_base::~_Function_base() /usr/local/twix-gcc-13/include/c++/13.2.0/bits/std_function.h:243:11
#17 0x0000000001d3bf93 std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>::~function() /usr/local/twix-gcc-13/include/c++/13.2.0/bits/std_function.h:334:11
#18 0x0000000001d3bf93 llvm::lto::LTO::run(std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, std::function<llvm::Expected<std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>> (unsigned int, llvm::StringRef, llvm::Twine const&)>) /twix/packages/twix-clang/work/llvm-project/llvm/lib/LTO/LTO.cpp:1191:41
#19 0x00000000011de452 lld::elf::BitcodeCompiler::compile() /twix/packages/twix-clang/work/llvm-project/lld/ELF/LTO.cpp:341:15
#20 0x00000000010dceef __gnu_cxx::__normal_iterator<lld::elf::InputFile**, std::vector<lld::elf::InputFile*, std::allocator<lld::elf::InputFile*>>>::__normal_iterator(lld::elf::InputFile** const&) /usr/local/twix-gcc-13/include/c++/13.2.0/bits/stl_iterator.h:1077:9
#21 0x00000000010dceef std::vector<lld::elf::InputFile*, std::allocator<lld::elf::InputFile*>>::begin() /usr/local/twix-gcc-13/include/c++/13.2.0/bits/stl_vector.h:871:16
#22 0x00000000010dceef void lld::elf::LinkerDriver::compileBitcodeFiles<llvm::object::ELFType<(llvm::support::endianness)1, true>>(bool) /twix/packages/twix-clang/work/llvm-project/lld/ELF/Driver.cpp:2343:3
#23 0x0000000001103b80 lld::elf::LinkerDriver::link(llvm::opt::InputArgList&) /twix/packages/twix-clang/work/llvm-project/lld/ELF/Driver.cpp:2810:3
#24 0x000000000110619d llvm::TimeTraceScope::~TimeTraceScope() /twix/packages/twix-clang/work/llvm-project/llvm/include/llvm/Support/TimeProfiler.h:155:37
#25 0x000000000110619d lld::elf::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) /twix/packages/twix-clang/work/llvm-project/lld/ELF/Driver.cpp:662:3
#26 0x0000000001107bfe lld::elf::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) /twix/packages/twix-clang/work/llvm-project/lld/ELF/Driver.cpp:157:20
#27 0x0000000000fccc3d lld::unsafeLldMain(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, llvm::ArrayRef<lld::DriverDef>, bool) /twix/packages/twix-clang/work/llvm-project/lld/Common/DriverDispatcher.cpp:164:11
#28 0x0000000000f1122b lld_main(int, char**, llvm::ToolContext const&) /twix/packages/twix-clang/work/llvm-project/lld/tools/lld/lld.cpp:93:27
#29 0x0000000000f11481 main /twix/packages/twix-clang/work/stage1/tools/lld/tools/lld/lld-driver.cpp:16:1
#30 0x00007fc94c9f4083 __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:342:3
#31 0x0000000000f10aae _start (/twix/packages/twix-clang/work/stage1/install/bin/ld.lld+0xf10aae)
```

**Comment by vient** [comment]
Removed unneeded parts from reproducer, confirmed that it reproduces on 17.0.1
Reproduces only with LTO=Full but not with Thin or Off.

Save this to `repro_64929.sh` and run `docker run -it -v $PWD:/test ubuntu:23.04 /test/repro_64929.sh`

```bash
#!/bin/bash -eux

apt update
apt install --yes git apt-transport-https build-essential cmake coreutils curl libc6-dev-i386 linux-tools-generic psmisc python3 python3-pip sed software-properties-common vim wget
python3 -m pip install --break-system-packages ninja

LLVM_GIT_BRANCH="llvmorg-17.0.1"

TOPLEV="$(realpath .)"
LLVM_PROJECT_DIR="${TOPLEV}/llvm-project"
BASE_COMPILER_DIR="${TOPLEV}/stage1"
BINUTILS_INCLUDE_DIR="${TOPLEV}/binutils/include"

mkdir -p "${TOPLEV}"
cd "$_"

git clone --depth 1 git://sourceware.org/git/binutils-gdb.git binutils
git clone -b "${LLVM_GIT_BRANCH}" --depth=1 https://github.com/llvm/llvm-project.git

# build base compiler
mkdir -p "${BASE_COMPILER_DIR}"
cd "$_"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_INSTALL_PREFIX="${BASE_COMPILER_DIR}/install" \
    -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld" \
    -DLLVM_ENABLE_RUNTIMES="compiler-rt" \
    -DLLVM_TARGETS_TO_BUILD=X86
ninja install
ninja -t clean

# build self-profiling compiler
mkdir -p "${TOPLEV}/stage2-prof-gen"
cd "$_"
CPATH="${BASE_COMPILER_DIR}/install/bin"
LLVM_EXE_LINKER_FLAGS="-fuse-ld=lld"
LLVM_SHARED_LINKER_FLAGS="-fuse-ld=lld"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_COMPILER="${CPATH}/clang" \
    -DCMAKE_C_FLAGS="-Xclang -mllvm -Xclang -vp-counters-per-site=256" \
    -DCMAKE_CXX_COMPILER="${CPATH}/clang++" \
    -DCMAKE_CXX_FLAGS="-Xclang -mllvm -Xclang -vp-counters-per-site=256" \
    -DCMAKE_EXE_LINKER_FLAGS="${LLVM_EXE_LINKER_FLAGS}" \
    -DCMAKE_SHARED_LINKER_FLAGS="${LLVM_SHARED_LINKER_FLAGS}" \
    -DCMAKE_INSTALL_PREFIX="${TOPLEV}/stage2-prof-gen/install" \
    -DLLVM_BINUTILS_INCDIR="${BINUTILS_INCLUDE_DIR}" \
    -DLLVM_BUILD_INSTRUMENTED=IR \
    -DLLVM_ENABLE_PROJECTS="clang;lld" \
    -DLLVM_PROFILE_DATA_DIR="${TOPLEV}/stage2-prof-gen/profiles/%p" \
    -DLLVM_TARGETS_TO_BUILD=X86 \
    -DLLVM_USE_LINKER=lld
ninja install
ninja -t clean

# gather compiler profiles
CPATH="${TOPLEV}/stage2-prof-gen/install/bin"

mkdir -p "${TOPLEV}/stage3-train-llvm-release"
cd "$_"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_COMPILER="${CPATH}/clang" \
    -DCMAKE_CXX_COMPILER="${CPATH}/clang++" \
    -DCMAKE_INSTALL_PREFIX="${PWD}/install" \
    -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra" \
    -DLLVM_TARGETS_TO_BUILD=X86 \
    -DLLVM_USE_LINKER=lld
ninja bin/llvm-addr2line  # random small target

rm -rf "${TOPLEV}/stage3-train"*

# merge profiles
cd "${TOPLEV}/stage2-prof-gen/profiles"
"${BASE_COMPILER_DIR}/install/bin/llvm-profdata" merge -output=clang.profdata ./*

# use PGO
mkdir -p "${TOPLEV}/stage2-prof-use-lto"
cd "$_"
CPATH="${BASE_COMPILER_DIR}/install/bin"
cmake -G Ninja "${LLVM_PROJECT_DIR}/llvm" \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_C_COMPILER="${CPATH}/clang" \
    -DCMAKE_CXX_COMPILER="${CPATH}/clang++" \
    -DCMAKE_INSTALL_PREFIX="${TOPLEV}/stage2-prof-use-lto/install" \
    -DLLVM_BINUTILS_INCDIR="${BINUTILS_INCLUDE_DIR}" \
    -DLLVM_ENABLE_LTO=Full \
    -DLLVM_ENABLE_PROJECTS="clang;lld" \
    -DLLVM_PROFDATA_FILE="${TOPLEV}/stage2-prof-gen/profiles/clang.profdata" \
    -DLLVM_TARGETS_TO_BUILD=X86 \
    -DLLVM_USE_LINKER=lld
ninja bin/dsymutil  # fails here
```

**Comment by vient** [comment]
Fixed somewhere between 17.0.4 and 17.0.5

**Comment by fhahn** [comment]
@vient  Thanks for checking with the latest released version and for closing!
